src/middlewared/middlewared/api/v25_04_0/app.py

```diff
@@ -1,13 +1,376 @@
+    'AppCertificateChoicesArgs', 'AppCertificateChoicesResult', 'AppCertificateAuthorityArgs',
+    'AppCertificateAuthorityResult', 'AppUsedPortsArgs', 'AppUsedPortsResult', 'AppIPChoicesArgs', 'AppIPChoicesResult',
+    'AppAvailableSpaceArgs', 'AppAvailableSpaceResult', 'AppGpuChoicesArgs', 'AppGpuChoicesResult', 'AppRollbackArgs',
+    'AppRollbackResult', 'AppRollbackVersionsArgs', 'AppRollbackVersionsResult', 'AppUpgradeArgs', 'AppUpgradeResult',
+    'AppUpgradeSummaryArgs', 'AppUpgradeSummaryResult',
 ]
 
 
+CONTAINER_STATES: TypeAlias = Literal['crashed', 'created', 'exited', 'running', 'starting']
+
+
+class HostPorts(BaseModel):
+    host_port: int
+    host_ip: str
+
+
+class UsedPorts(BaseModel):
+    container_port: int
+    protocol: str
+    host_ports: list[HostPorts]
+
+
+class AppVolumes(BaseModel):
+    source: str
+    destination: str
+    mode: str
+    type_: str = Field(alias='type')
+
+
+class AppContainerDetails(BaseModel):
+    id: str
+    service_name: str
+    image: str
+    port_config: list[UsedPorts]
+    state: CONTAINER_STATES
+    volume_mounts: list[AppVolumes]
+
+
+class AppNetworks(BaseModel):
+    Name: str
+    Id: str
+    Labels: dict
+
+    model_config = ConfigDict(extra='allow')
+
+
+class AppActiveWorkloads(BaseModel):
+    containers: int
+    used_ports: list[UsedPorts]
+    container_details: list[AppContainerDetails]
+    volumes: list[AppVolumes]
+    images: list[NonEmptyString]
+    networks: list[AppNetworks]
+
+
+class AppEntry(BaseModel):
+    name: NonEmptyString
+    id: NonEmptyString
+    state: Literal['CRASHED', 'DEPLOYING', 'RUNNING', 'STOPPED', 'STOPPING']
+    upgrade_available: bool
+    latest_version: NonEmptyString | None
+    image_updates_available: bool
+    custom_app: bool
+    migrated: bool
+    human_version: NonEmptyString
+    version: NonEmptyString
+    metadata: dict
+    active_workloads: AppActiveWorkloads
+    notes: LongString
+    portals: dict
+    version_details: dict | None = None
+
+
+@single_argument_args('app_create')
+class AppCreateArgs(BaseModel):
+    custom_app: bool = False
+    values: Secret[dict] = Field(default_factory=dict)
+    custom_compose_config: Secret[dict] = Field(default_factory=dict)
+    custom_compose_config_string: Secret[LongString] = ''
+    catalog_app: str | None = None
+    app_name: str = Field(pattern=r'^[a-z]([-a-z0-9]*[a-z0-9])?$', min_length=1, max_length=40)
+    '''
+    Application name must have the following:
+    1) Lowercase alphanumeric characters can be specified
+    2) Name must start with an alphabetic character and can end with alphanumeric character
+    3) Hyphen '-' is allowed but not as the first or last character
+    e.g abc123, abc, abcd-1232
+    '''
+    train: NonEmptyString = 'stable'
+    version: NonEmptyString = 'latest'
+
+
+class AppCreateResult(BaseModel):
+    result: AppEntry
+
+
+class AppUpdate(BaseModel):
```

Comment:
No `ForUpdateMetaclass`? Setting defaults here may clear out the fields not provided in the call params depending on how `app.update` updates the fields.

Reply:
`app.update` expects the fields and this is keeping the behavior which was there already. We probably will have to tweak the implementation a bit otherwise. If you think i should do that, let me know

Reply:
Ok, I was just checking because I know our update endpoints don't all work the same.